### PR TITLE
[RETRACTED]; compressed-memory: available only via patches for projects that actively use it

### PR DESCRIPTION
**Pull request by NoTengoBattery**
[RETRACTED]

Even when rejected by OpenWrt, this pull request is being actively used by several projects and builds. They include my own builds and builds from other users/brands with a wider spread. Even when retracted, these projects will still benefit from the changes by pulling the latest changes by comparing my master branch against OpenWrt master.

This means that this project will work, always, against the OpenWrt master branch. I'll keep it updated, but you may find it unreliable with older versions. I do not recommend relying on this pull request anymore. It is officially retracted now. All developers should pull the changes from my master branch or directly from the link below.

More information available here: https://forum.openwrt.org/t/49076

### CODE REVIEW

**Comment by bobafetthotmail** [comment]
I think the PR (and the title) should be 100% about adding `compressed-memory` package, deprecating or removing zram-swap can be done or discussed later.

If the two packages must coexist in the repository I think you should add a negated dependency ( **!zram-swap** I guess? How does this work in OpenWrt) so that you can't install both this or its kmods and zram-swap at the same time.

That said, I like this.

**Comment by NoTengoBattery** [comment]
Well, no.

They cannot coexist torgerther and the negative dependency is already in the commits.

This commits does not remove `zram-swap` either. The title does not mean "remove OpenWrt's zram in favour of my stuff", it only means "make it obsolete by providing a way better alternative". None of the commits talk about that, only the title because I wanted to keep it tidy for the ones that get a first glance without actually reading the commits (like you, but that's fine, thanks for bringing it up).

So, in fact, this commit is 100% about adding the package with a negative dependency which deprecates but does not removes the current alternative.

**Comment by NoTengoBattery** [comment]
Just one thing to add: this package is already rested and does not require work to be done. The LuCI app is ready, and the pull request will be submitted to the LuCI repository upong this getting merged (if at all).

File: package/system/compressed-memory/files/zram.init [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+PROG="/bin/zram"
+EXTRA_COMMANDS="stats compact"
+
+start_service() {
+	procd_open_instance "zram"
+	procd_set_param command "$PROG"
+	procd_append_param command start
+	procd_set_param env RUNNABLE=$RUNNABLE ZSWAP_OK=$ZSWAP_OK
+	procd_set_param stdout 0
+	procd_set_param stderr 0
```
Comment (hauke) [inline package/system/compressed-memory/files/zram.init]: I think explicitly setting ```procd_set_param stderr 0``` is not needed

File: package/kernel/linux/modules/crypto.mk [hunk 2]
```diff
@@ -809,3 +809,18 @@ endef
...
 $(eval $(call KernelPackage,crypto-xts))
 
+
+define KernelPackage/crypto-zstd
+  TITLE:=zstd compression CryptoAPI module
+  DEPENDS:=+kmod-lib-zstd +kmod-crypto-acompress
+  KCONFIG:=CONFIG_CRYPTO_ZSTD
+  FILES:=$(LINUX_DIR)/crypto/zstd.ko
+  AUTOLOAD:=$(call AutoLoad,09,zstd)
+  $(call AddDepends/crypto)
```
[the remark this answers is not in the document]
Comment (NoTengoBattery) [reply package/kernel/linux/modules/crypto.mk]: I'm gonna test it soon.

File: package/kernel/linux/modules/other.mk [hunk 3]
```diff
@@ -869,6 +869,85 @@ endef
...
+define KernelPackage/zswap
+  SUBMENU:=$(OTHER_MENU)
+  TITLE:=zswap compressed swapping cache
+  DEPENDS:=+kmod-zsmalloc
+  KCONFIG:= \
+	CONFIG_FRONTSWAP=y \
+	CONFIG_Z3FOLD \
+	CONFIG_ZBUD \
+	CONFIG_ZPOOL \
+	CONFIG_ZSWAP=y
```
[the remark this answers is not in the document]
Comment (NoTengoBattery) [reply package/kernel/linux/modules/other.mk]: zswap is not a module and therefore there is no solution about that problem. You don't have to believe me, explore the kernel source code. The same is true for frontswap. The memory allocators and the zpool are indeed modules. If you think about a solution (removing zswap is not a solution, because it doesn't solve the problem, just to clarify), I'll like to know it.

**Comment by CodeFetch** [comment]
Why can't you create separate packages for zram and zswap? These are two different things...

**Comment by NoTengoBattery** [comment]
I'm not going to answer that question because they are different packages
already and makes no sense at all to lose my time explaining that when the
source code is crystal clear.
Thanks for your question and I hope is clear now.

File: package/kernel/linux/modules/other.mk (hunk 3 above)
[the remark this answers is not in the document]
Comment (NoTengoBattery) [reply package/kernel/linux/modules/other.mk]: An interesting thing about the build bot is that this package is in conflict with `zram-swap`, and it will never be built automatically and your comment, that the kernel will always build the `zswap` code is, therefore, false.

The code is designed to *eventually* totally eradicate the `zram-swap` package because it is so bad in quality that I really doubt how it is still in the official repositories.

And I'm not judging for free the package, I did created a way better solution ready to end the life of `zram-swap`.

File: package/kernel/linux/modules/crypto.mk (hunk 2 above)
[the remark this answers is not in the document]
Comment (NoTengoBattery) [reply package/kernel/linux/modules/crypto.mk]: About this: modules do work with `AutoProbe`, but as long as the default OpenWrt packages use `AutoLoad`, it makes no sense at all. Mostly because the new kernel packages depend on the original packages and mixing the both options does not seems logical, unless someone can provide me a link to a official announcement of refactoring.

**Comment by CodeFetch** [comment]
@NoTengoBattery I see the source code, but not why it should be clear. Actually your commit descriptions are vague and your package names are misleading. zram != zswap...

File: package/kernel/linux/modules/other.mk (hunk 3 above)
[the remark this answers is not in the document]
Comment (ynezz) [reply package/kernel/linux/modules/other.mk]: > zswap is not a module and therefore there is no solution about that problem.

This features increases resulting 5.4.28 zImage by 7344 bytes on imx6/cortex-a9 so I guess, that it's going to be similar on other targets.

I'll put aside, that this is very niche feature usable only on a few targets/devices, but I think, that it's not acceptable to include this config options by default on targets constrained by the SMALL_FLASH constraint etc. IIRC most of those targets don't use swap at all, because they don't have suitable block device for use etc.

> zpool are indeed modules

zpool is also built-in once the zswap is built-in, means always.

**Comment by bobafetthotmail** [comment]
> This commits does not remove `zram-swap` either. The title does not mean "remove OpenWrt's zram in favour of my stuff", it only means "make it obsolete by providing a way better alternative".

PR titles, just like commit titles, should really be about what the PR is, you are adding a package, not deprecating or obsoleting anything. And my point still stands.
May not matter much in the end as it's not merged with the commit, but if you need to explain something that should be self-explanatory...

> wanted to keep it tidy for the ones that get a first glance without actually reading the commits

Weird, because by reading the PR I thought you actually removed that package and only after I read the commits I saw what you actually did.

**Comment by bobafetthotmail** [comment]
> @NoTengoBattery I see the source code, but not why it should be clear. Actually your commit descriptions are vague and your package names are misleading. zram != zswap...

While I dislike his attitude, the source is indeed kind of obvious.

If you see the commit https://github.com/openwrt/openwrt/pull/2840/commits/6275cf262846505efc400ba778b80d1550d51ff0
he is adding 

1. a generic package called **compressed-memory** that is just uci-related config and is only a dependency https://github.com/openwrt/openwrt/pull/2840/commits/6275cf262846505efc400ba778b80d1550d51ff0#diff-56f750fadd27bdadf424761433b31950R17
2. a package called **swap-zram-writeback**  that has the script and depends on tools and kmods for zram https://github.com/openwrt/openwrt/pull/2840/commits/6275cf262846505efc400ba778b80d1550d51ff0#diff-56f750fadd27bdadf424761433b31950R39
3. a package called **swap-zswap-cache** that has the script and depends on tools and kmods for zswap https://github.com/openwrt/openwrt/pull/2840/commits/6275cf262846505efc400ba778b80d1550d51ff0#diff-56f750fadd27bdadf424761433b31950R60

File: package/kernel/linux/modules/other.mk (hunk 3 above)
[the remark this answers is not in the document]
Comment (NoTengoBattery) [reply package/kernel/linux/modules/other.mk]: Is this the only change needed?
Because if we talk about the default config, the packages will never be built by default as long as `zram-swap` exist. But I can trivially add the negative dependency, not a big deal. The packages are idependent (not as @CodeFetch somehow thinks) and without `zswap`, it will only use the advanced features for `zram`.

As a side note: please note that the new `zram` enables `lz4hc` and `zstd`, so for SMALL_FLASH they need somehow space to exist.

Let me know what else is needed for this to be merged.

**Comment by NoTengoBattery** [comment]
I sorry, @bobafetthotmail, but Cambridge is not in line with your thinking: [deprecate](https://dictionary.cambridge.org/dictionary/english/deprecate). It didn't, don't, and won't ever mean remove. It's okay to be confused, but I will not change my title because it's not okay to think that you have a point if you only give me your personal experience as a proof.

As long as it have a negative dependecy on the original package, the title is correct. Because it does not just "install a new package", it does deselect automatically `zram-swap`, and that is the basic meaning of `deprecation`.

---

If you are not reading the source code, and the commit, and like @CodeFetch, you come here to say thinks that are clearly not true, you are confusing other developers that only come to barely see anything. And I am not into exlaining them anything because whoever will merge this **will have to read the whole source code and commits**, and the opinion of other people that just read 2 or three lines like the two of you, is only a noisy opinion.

**Comment by adrianschmutzler** [comment]
@NoTengoBattery Please watch your tone and attitude.
You have received feedback on the content and I would appreciate if you could limit your answers to that as well, instead of having 80 % of your comments about how others should review your contribution.
Note that by responding to feedback aggressively (as you did repeatedly), you just weaken the credibility of what you present.

File: package/kernel/linux/modules/other.mk (hunk 3 above)
[the remark this answers is not in the document]
Comment (NoTengoBattery) [reply package/kernel/linux/modules/other.mk]: > I'll put aside, that this is very niche feature usable only on a few targets/devices, but I think, that it's not acceptable to include this config options by default on targets constrained by the SMALL_FLASH constraint etc. IIRC most of those targets don't use swap at all, because they don't have suitable block device for use etc.

I have something to add here, @ynezz:
The idea behind using `zswap` is to be a cache for `zswap`. You don't need a backing storage and that's why they share a configuration: they are meant to be used torgerther.
Why? Let me explain:
1. `zswap` is a cache, while `zram` is a compressed RAM disk.
2. Because of how the `zpool` works compared to how the compressed block device work, `zswap` *is faster*.
3. The idea is to use the faster `zswap` in front (that's why the kernel feature is called FRONTSWAP) is to save time compared to the time required to do the same work in `zram`, which is a direct winning in performance.
4. In the back of `zswap` is `zram` with a slower but does a way better job at compressing, so we have a direct win on resource usage.
5. The two layers will provide the best of them, one being faster and the other using a better algorithm for compression.
6. That is true even if no backing storage is used. But the benefits are event greater if it exists because great amounts of IO will be saved and, when the memory cannot be compressed, great amounts of incomprensible unused data will be moved from memory to the backing storage, freeing even more memory.

So `zswap` is not meant to be used as a niche feature, but as direct a performance benefit. If small flash devices cannot run with the extra modules, I think it's better to disable this package completely instead of only disabling `zswap`.

**Comment by NoTengoBattery** [comment]
@adrianschmutzler you are right, but none of you are providing feedback.

@hauke and @yneez are providing feedback, you aren't.

**Comment by bobafetthotmail** [comment]
> I will not change my title because it's not okay to think that you have a point if you only give me your personal experience as a proof.

That's a fair request. "deprecate" has been already used in other commits in the tree to mean "remove", or for the very least "add deprecation notices" on the deprecated (but still available) package. You are not doing any of that, so I just found it a bit odd.
See some examples below:

https://github.com/openwrt/openwrt/commit/ed4ac0ed65175230d6a7d3196974e6967b31a074
https://github.com/openwrt/openwrt/commit/301baf34ddbc838b61dfd8564b27a440f0b85e5a
https://github.com/openwrt/openwrt/commit/9f63b7326f5cddb747ce2a9e15023df97f8db3a4
https://github.com/openwrt/openwrt/commit/c7c4ae35d6e9b53e2c66da7d6eadc9d77d64ed60
https://github.com/openwrt/openwrt/commit/899a23227e5587078e71b599ca447ccfb9b8af79
https://github.com/openwrt/openwrt/commit/335117b87c7f983d19b6e6ed0e579722f45911d7

>As long as it have a negative dependecy on the original package, the title is correct. Because it does not just "install a new package", it does deselect automatically zram-swap, and that is the basic meaning of deprecation.

Setting negative dependencies does not make something deprecated, it's just incompatible with each other. See above for examples of "deprecation" (adding notices or removing outright).

**Comment by adrianschmutzler** [comment]
> @adrianschmutzler you are right, but none of you are providing feedback.

Yes. I haven't provided any feedback. I've just followed the discussion and wanted to make sure this doesn't escalate, in order to maintain rule #12 (https://openwrt.org/rules).

@CodeFetch did provide feedback, and whether you liked that or not, my intention was that you should stick to the part where you answer to it based on its content, and not based on what you think it tells about the work attitude of its author.

**Comment by NoTengoBattery** [comment]
@adrianschmutzler I agree. I will give up my own comments, they are unrelated because the "feedback" provided is also unrelated. I shouldn't have answered anyway.

What I dislike is not criticism, I am accepting and working with it in the source code. But tell me, what kind of feedback is when people didn't ever read the source code and commend things that are not real problems? Just noisy opinions.

**Comment by NoTengoBattery** [comment]
@CodeFetch:
Good question, I didn't made it that way because:
1. Even if they can be used separate, they are meant to be used torgerther.
2. They both conflict and depend on the same packages, making them "separate" would end up with code duplication for the dependencies.

**Comment by CodeFetch** [comment]
@NoTengoBattery Thank you for your answer. I'll have a look at the documentation to understand why they can be used together. I assumed that zram and zswap should not be used together as they basically provide the same functionality while zswap can use another block-device to keep cold pages, but has a little more overhead.

>     2. They both conflict and depend on the same packages, making them "separate" would end up with code duplication for the dependencies.

Hm... That sounds to me like they can't be used together on the other hand... In embedded development code duplication in the source code can be preferred over "dead code" in the resulting firmware. That's why I'd create two different packages if it saves a few bytes and outsource common code into a library. (That's just my rule of thumb - In this case I agree it's nit-picking). I'll have a deeper look... We actually use zram-swap and I'm quite happy with it, but this might be better. Thank you for your contribution!

**Comment by NoTengoBattery** [comment]
Oh yeah, I understand the point but the `small flash` feature, that changes everything, see: for small flash devices, `zram-swap` **is** the natural choice. But for more powerful devices where latency changes are more noticeable, and savings are more needed (because people tend to use them as NAS or ad-blockers) it is meaningful when it comes to come extra kb. So disabling the whole feature for small flash devices and deprecating `zram-swap` in favour of this for all other targets can be really useful.

And I see the confusion about "they do the same", but they don't. `zram` is a compressed RAM disk while `zswap` is a RLU cache. `zswap`, due the `zpool` architecture is faster (in theory) than `zram`. But `zram` due to it's block nature can compress way more than the other. Testing has shown that while `zswap` can be blazing fast, being optimistic is can only reach 2.80 in compression ratio, and it's limited to 3.00. `zram` is slower but can reach as high as 5.50 in compression ratio, without limit in the compression.

So, the idea is to use the fastest cache to allow the kernel to benefit from the faster memory allocation, combined with `lz4`, the fastest compressor, for the RLU mechanics, which means that pages that are use often will save space while not giving up performance (even gaining performance compared to `zram` alone).

But the use of `zram` is to back up the operation of the cache. When the cache fills or pages are not accessed (most startup and library code will only be loaded, modified and kept around doing nothing for a long time), the block nature of `zram` with a slower but better compression algorithm (like `zstd` or even `deflate`) can save even more memory for pages rarely used.

That brings up the best of both worlds: a blazing fast cache that saves upto 2/3 of the memory for pages being used often plus a space for pages not used often that can save upto 4/5 of the memory they consume.

And when adding a real storage (like devices with USB ports), the new feature of `zram`, the "backing storage" will make `zram` a kind of L2 cache for the disk, saving I/O at all costs when using what I call the "triple layer" of memory management, and removing never used and incompressible pages from RAM, freeing even more memory in the end.

So yes, this doesn't look good for powerless devices, because it's designed for powerful devices that may need some help (like the IPQ4018 which is quad core with USB 3.0; but with only 256 MiB of RAM). As I said before, the small flash feature which I didn't know changes the game.

If you can, @hauke and @ynezz, please read this more detailed explanation. This is also discussed here, if you can take a look at it:
https://github.com/NoTengoBattery/openwrt/blob/26fe1a8c86045968c8aee8366648f10b4bfd84fd/notes-final.txt#L155

---
PS.: Sorry for being rude. Got a bad new today, wasn't in my best. I'm not like that, I only want to help OpenWrt because I love it. Forgive me, if you can, but that will never happen again.

**Comment by ynezz** [comment]
tl;dr sorry :-)

> the small flash feature which I didn't know changes the game

What about starting low profile and making this optional for the start? I would simply add a config option in `config/Config-kernel.in` which would be disabled by default (not adding `default y`) and add proper dependencies. As of 5.6 `zswap` is still marked as `(EXPERIMENTAL)` in upstream. Once tested it could be selectively enabled by default on those targets where it makes sense and works.

**Comment by CodeFetch** [comment]
@NoTengoBattery 
> If you can, @hauke and @ynezz, please read this more detailed explanation. This is also discussed here, if you can take a look at it:
> https://github.com/NoTengoBattery/openwrt/blob/26fe1a8c86045968c8aee8366648f10b4bfd84fd/notes-final.txt#L155

Very comprehensive documentation. I'd like to see this in the Wiki... A link to this would have saved me from asking questions and you from the stress of reading them. I know it's sometimes hard to understand why people ask such questions, but we are all specialists here and a thorough explanation saves people from diving too deep into one others specialty. I'm a low-level programmer and know much of the linux source code like the wireless and network stack and file systems, but I don't have a clue about the page compression magic here. Actually I think you are 1 of < 100 on this planet who knows... Don't take your knowledge for granted, but share it. That's the purpose of open-source development.

I'll definitely give it a try.

**Comment by NoTengoBattery** [comment]
@ynezz ok, `zswap` is experimental even if it's mature and it's been in the kernel for more than 6 years and widely used by Sony (they indeed authored `z3fold`), they are sometimes weird in upstream. But however... it is optional anyway, the source code works with or without it.

But the real deal is `zram`. It is a widely used feature, but my source code is in natural conflict with OpenWrt's one. And as you won't accept the removal of your `zram`, they must conflict. And this is a problem because if they conflict and you choose your `zram` for the bot, this will never ever see a light. Because the LuCI app cannot be built without it's dependencies here (and I'm waiting for this to be merged to go LuCI and make the pull request), the app will also lose all of it's value.

So, my proposal is to use this for the non low-profile devices by disabling it for the small flash ones, and letting in the `zswap` part in the source code, but disabled and available for those who want to build custom ROMs, as `zswap` in combination with `zram` is widely used in the Android modding world (and some official OEM ROMs, such as Xperia ROMs) for the reasons explained here.

Yes, I was an Android modder in the good'ol days of XDA Developers, so I have a perspective about this that probably you don't. However, the `zram` part and the LuCI app is worth even if you want to be conservative about `zswap`. Let me know what we can do to merge this feature into OpenWrt, I'm open to ideas and ready to work on it.

P.S. only the kernel dependencies are a problem because the scripts and the app are tested and ready, but if you see a small bug or a feature request, I'm also into it.

**Comment by r-salvaterra** [comment]
There's no facility (yet, I hope) in Linux which supports paging out compressed data to a swap partition, unfortunately (as this would come with a nice I/O reduction). Pages in the zswap pool are always decompressed when written out, either because they're uncompressible or because the pool is full. Likewise, when specifying a backing swap device for zram, only uncompressible pages are swapped out until the zram device is full, after which all pages are sent uncompressed to the backing device.

When combining zswap with zram, we incur in the overhead of having to decompress the page from the zswap cache, only to recompress it again in zram. Additionally, we're increasing the number of allocations/deallocations required under memory pressure, something we really want to avoid.

For these reasons and for what its worth, I have to heartily recommend against pulling this.

**Comment by NoTengoBattery** [comment]
That's not precisely true. You are totally forgetting that the zram compression algorithm is better than the zswap on the default configuration.

Under your argument, then having 2 or 3 levels of cache in a CPU will make no sense because "we always incur in overhead of dropping cache lines when reading from memory, specially when the L1 is full" and that's absurd. That's why zswap *is a cache* and zram with write back becomes effectively *a cache*, so you totally missed the point.

The extra allocations are absorbed when zram evicts pages to the physical backing device, a detail that you clearly missed because when you use zram as default in OpenWrt the pages remain in memory forever and this is not the case here. This is not designed to solve memory problems, this is designed with a 2 cache-level architecture for workloads that exhaust the memory but doesn't use that memory every nanosecond. If you understand the concept behind zram, anyway, you understand that zram itself is not the magic key to performance and works better for a particular workload only.

And if it wasn't enough every feature is optional. You are not forced to enable zram or zswap, and the configuration is resistant to whatever configuration you throw it.

Also, @r-salvaterra, check the dates before commenting because this is pretty much dead. Nobody is going to pull this but thank you for your "healthy" recommendations.
